KAFKA-15038: Add metadatacache into RemoteLogManager, and refactor all relevant codes

Owen-CH-Leung opened the pull request:
PR in response to [KAFKA-15038](https://issues.apache.org/jira/browse/KAFKA-15038). This PR incorporates `MetaDataCache` into `RemoteLogManager` and refactor all other relevant codes
 
### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

Owen-CH-Leung committed:
Add metadatacache into RemoteLogManager, and refactor all relevant codes

Owen-CH-Leung commented:
Hello @divijvaidya , I'm still in the process of writing the PR and would like to ask something : 

For this ticket, in order to reuse the metadata cache inside broker (which I believe is already instantiated when the broker server is set up), I think I need to access the following factory method code snippet : 

https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/MetadataCache.scala#L114-L125

However, this method will always return a new instance of metadatacache instead of the one being instantiated inside broker. So does it make sense to change it to a Singleton class ?

divijvaidya commented:
I am not sure I understand why we need to access the factory method. 

The metadataCache instance is created on broker startup. It is then passed to ReplicaManager. RemoteLogManager is also created on start-up and you can pass this metadataCache in the constructor of RemoteLogManager at https://github.com/apache/kafka/blob/8aaf7daff393f2b26438fb7fe28016e06e23558c/core/src/main/scala/kafka/server/BrokerServer.scala#L573 . I observed that you are already doing this in the change in this PR. 

Once we have an instance inside RemoteLogManager, you can call Zk/Kraft specific functions using: `val zkMetadataCache = metadataCache.asInstanceOf[ZkMetadataCache]`

Owen-CH-Leung commented:
Ahh yes..Sorry for the dumb question!

I was getting really confused by the test result. From the CI failure I saw there're 2 failed test, which are `ListOffsetsRequestTest > testResponseDefaultOffsetAndLeaderEpochForAllVersions()` and `ListOffsetsRequestTest > testResponseIncludesLeaderEpoch()` respectively.

when I executed the tests locally via `./gradlew core:test`, I can see that these 2 tests have assertion failure also (I executed it multiple times and none of them are success). But when I execute only the specfic test suite like `./gradlew core:test --tests "kafka.server.ListOffsetsRequestTest"`, sometimes these 2 tests would fail and sometimes it succeeds...

Is there some setup difference when executing `./gradlew core:test` vs `./gradlew core:test --tests "kafka.server.ListOffsetsRequestTest"`  ? 

<img width="1167" alt="image" src="https://github.com/apache/kafka/assets/43698890/c81a44c5-28c1-47d5-b429-3391b0eb63c9">

Owen-CH-Leung committed:
Fix failing CI jobs

Owen-CH-Leung committed:
Remove commented code

Owen-CH-Leung commented:
Oh yes...Thanks and sorry for the dumb question!

I've just managed to fix the failing CI test from the previous run, but it seems that it's now getting error at git checkout...Can I ask for your help to re-start the CI job ?

Owen-CH-Leung committed:
Fix formatting

Owen-CH-Leung commented:
Oh yes....Thanks and sorry for the dumb question! 

I've fixed the CI failing test and all should be good for review now. Let me know your feedback.

divijvaidya commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -144,8 +145,9 @@ public class RemoteLogManager implements Closeable {
-    // topic ids that are received on leadership changes, this map is cleared on stop partitions
-    private final ConcurrentMap<TopicPartition, Uuid> topicPartitionIds = new ConcurrentHashMap<>();
+    // topic ids that are received on leadership changes & on stop partitions are stored in this cache
+    private final MetadataCache metaDataCache;
```
nit

s\metaData\metadata

(everywhere)

Owen-CH-Leung replied:
Got it. Will revise accordingly

Owen-CH-Leung committed:
Merge branch 'trunk' into KAFKA-15038

Owen-CH-Leung committed:
Fix wrong merge trunk which result in code loss

divijvaidya commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -347,15 +353,16 @@ public void onLeadershipChange(Set<Partition> partitionsBecomeLeader,
-            Uuid topicIdPartition = topicPartitionIds.remove(topicPartition);
-            LOGGER.debug("Removed partition: {} from topicPartitionIds", topicIdPartition);
+            Map<String, Uuid> mapping = metaDataCache.topicNamesToIds();
+            Uuid topicIdPartition = mapping.remove(topicPartition.topic());
```
The idea of replacing this cache with Metadata cache is that the existing metadata cache will serve as the source of truth. It should already be updated to contains the correct topics Id <-> topicPartition mapping irrespective of whether tiered storage is turned on or not. TS code should only access this cache and not modify it in any way. The modification (adding/removal) should already exist is rest of the normal path Kafka code. (If it doesn't then we need to fix it).

Owen-CH-Leung replied:
Got it. So can I say for `stopPartition` and `onLeadershipChange`, there's no need to update the metadatacache at all ? Is there any extra logic that should be addded into these 2 APIs ?

clolov replied:
I don't think there is anything to be done here. The topic partitions should disappear once the controller lets us know they should.

Owen-CH-Leung replied:
@clolov Thanks a lot for your comment. I've incorporated your comments

clolov commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -280,13 +285,18 @@ public RemoteStorageManager storageManager() {
 
     private void cacheTopicPartitionIds(TopicIdPartition topicIdPartition) {
-        Uuid previousTopicId = topicPartitionIds.put(topicIdPartition.topicPartition(), topicIdPartition.topicId());
+        Map<String, Uuid> mapping = metaDataCache.topicNamesToIds();
```
Adding to Divij's comment - this is an unmodifiable map. Even if you wanted to you wouldn't be able to add to it:
```
  def topicNamesToIds(): util.Map[String, Uuid] = {
    Collections.unmodifiableMap(metadataSnapshot.topicIds.asJava)
  }
```

Owen-CH-Leung replied:
Thanks. I've removed this logic

clolov reviewed (commented):
Carried out a first review and left a couple of comments.

Owen-CH-Leung committed:
Code refactor, Remove metadatacache updating logic

Owen-CH-Leung committed:
Merge branch 'trunk' into KAFKA-15038

Owen-CH-Leung committed:
Resolve code conflict

clolov reviewed (approved):
Great, thank you! I believe this looks good to me!

clolov commented:
To provide some clarity, I will just do one more pass for the MetadataCache itself, because I would like to confirm that it is updated when we would expect it to be (as per Divij's comment)

kamalcph commented on core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java:
```diff
@@ -802,41 +817,25 @@ private void verifyInCache(TopicIdPartition... topicIdPartitions) {
-    }
-
     @Test
     void testTopicIdCacheUpdates() throws RemoteStorageException {
```
Let's remove this test.

Owen-CH-Leung replied:
Sure. Removed the test

kamalcph commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -3600,7 +3600,10 @@ class ReplicaManagerTest {
       _ => Optional.of(mockLog),
-      brokerTopicStats)
+      brokerTopicStats,
+      mock(classOf[MetadataCache])
```
Could you please extract the `mock` as variable before using it to keep it uniform?

```
val metadataCache =  mock(classOf[MetadataCache])
...
when(metadataCache.getTopicId(topic))...

```

Owen-CH-Leung replied:
Sure. Moved the `mock(classOf[MetadataCache])` as variable now.

kamalcph reviewed (commented):
Overall LGTM. Can we hold this patch until #13947 lands? Otherwise, this patch needs one integration test to verify end-to-end change.

Owen-CH-Leung committed:
Remove test & small code refactor

Owen-CH-Leung commented:
> Overall LGTM. Can we hold this patch until #13947 lands? Otherwise, this patch needs one integration test to verify end-to-end change.

No problem. I can perform rebase after the PR was merged

kamalcph reviewed (approved):
LGTM

Owen-CH-Leung committed:
Merge branch 'trunk' into KAFKA-15038

Owen-CH-Leung committed:
refactor test suite

Owen-CH-Leung commented:
@clolov @kamalcph I've merged base and refactored the test suite a bit. There're a few failed test but I believe those are unrelated to the changes I made. Could you please take a look and let me know your comments ?

kamalcph reviewed (approved):
LGTM. 

@satishd Call fore review. PTAL.

showuon commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -378,7 +372,7 @@ public void stopPartitions(Set<TopicPartition> topicPartitions,
         if (delete) {
             // NOTE: this#stopPartitions method is called when Replica state changes to Offline and ReplicaDeletionStarted
-            topicPartitions.forEach(topicIdByPartitionMap::remove);
+//            topicPartitions.forEach(topicIdByPartitionMap::remove);
```
Should we delete this line directly?

Owen-CH-Leung replied:
Thanks. Removed.

showuon commented on core/src/main/java/kafka/log/remote/RemoteLogManager.java:
```diff
@@ -285,19 +288,15 @@ public RemoteStorageManager storageManager() {
         return remoteLogStorageManager;
     }
 
+    public MetadataCache metadataCache() {
+        return metadataCache;
```
Why do we need this? For test only? I think the test doesn't have to use it. See my comment below.

Owen-CH-Leung replied:
Yes agree that we don't really need this. I've removed it

showuon commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -3624,14 +3624,18 @@ class ReplicaManagerTest {
+      brokerTopicStats,
+      mockMetadataCache
+    )
+    when(remoteLogManager.metadataCache().getTopicId(topic)).thenReturn(topicId)
```
What's the difference between `mockMetadataCache` and `remoteLogManager.metadataCache()`? If no difference, could we use the former one and remove the `remoteLogManager#metadataCache()` method?

Owen-CH-Leung replied:
Thanks. I've changed to use `mockMetadataCache` instead

showuon reviewed (commented):
Thanks for the PR. Overall LGTM. Left some comments.

Owen-CH-Leung committed:
Remove unwanted code block

showuon reviewed (approved):
LGTM!

satishd reviewed (approved):
Thanks @Owen-CH-Leung for the PR. Overall LGTM.

satishd commented:
@divijvaidya Let me know if you still have any comments on this PR.  We can merge this PR if there are no blocking comments. 

There are a few unrelated test failures in jenkins jobs.

divijvaidya commented:
> @divijvaidya Let me know if you still have any comments on this PR. We can merge this PR if there are no blocking comments.
> 
> There are a few unrelated test failures in jenkins jobs.

I want to ensure that there are no race conditions between LISR and updateMetadata requests from controller i.e. validate that metadata cache is updated during LISR request. Give me a day to validate that.

showuon commented:
> I want to ensure that there are no race conditions between LISR and updateMetadata requests from controller i.e. validate that metadata cache is updated during LISR request. Give me a day to validate that.

This is a really great point. I've taken a look:
 - For KRaft, the metadata cache is synced up when `RLM#onLeadershipChange`.
 - For ZK, the controller will send out LISR and updateMetadataRequest basically at the same time (but LISR is enqueued first). I think in ZK, there might be chances the topicID is still not updated in metadataCache before `RLM#onLeadershipChange` is invoked. 
 
I think we need to have some solution for it, or we revert this change before ZK got removed in v4.0? Thoughts?

clolov commented:
I am possibly wrong, but I think that even though this piece of code sends LISR requests first 
https://github.com/apache/kafka/blob/88d2c4460a1c8c8cf5dbcc9edb43f42fe898ca00/core/src/main/scala/kafka/controller/ControllerChannelManager.scala#L720-L722 there are never requests of different kinds at the same time based on the creation of a new batch in these scenarios:
* https://github.com/apache/kafka/blob/88d2c4460a1c8c8cf5dbcc9edb43f42fe898ca00/core/src/main/scala/kafka/controller/KafkaController.scala#L1117-L1127
* https://github.com/apache/kafka/blob/88d2c4460a1c8c8cf5dbcc9edb43f42fe898ca00/core/src/main/scala/kafka/controller/KafkaController.scala#L1237-L1239
* https://github.com/apache/kafka/blob/88d2c4460a1c8c8cf5dbcc9edb43f42fe898ca00/core/src/main/scala/kafka/controller/KafkaController.scala#L1396-L1400

showuon commented:
@clolov , thanks for the correction. :)

So, I think we need to have some solution for the problem in ZK's mode (which LISR will be sent out first before updateMetadataRequest), or we keep this PR open before ZK got removed in v4.0? Thoughts?

For me, I think we keep this PR open, which is the simplest way.

clolov commented:
Sorry, I do not understand the concern. Let me explain my reasoning.

Based on https://github.com/apache/kafka/pull/14136#issuecomment-1691749594 the controller sends blocking requests in the order LISR, UpdateMetadata, StopReplica (which eventually trickle down to `stopPartitions`).

Let's consider the usage of the pseudo metadata cache (from now on referred to as the concurrent map) in the RemoteLogManager before the change proposed in this pull request.

`onLeadershipChange` does not lookup topic identifiers in the concurrent map, it only populates it with entries.
The places where topic identifiers are looked up are in `stopPartitions`, `fetchRemoteLogSegmentMetadata`, and `findOffsetByTimestamp`.

`stopPartitions` is only called last in the list of requests sent by the controller. This cannot arrive before the previous update metadata has propagated due to the blocking behaviour of the controller sending requests.

Both `fetchRemoteLogSegmentMetadata` and `findOffsetByTimestamp` are on the read path of Tiered Storage i.e. they are triggered by consumers. In order for consumers to continue consuming after a leadership change they need to get the latest metadata, as such update metadata requests need to have propagated through the cluster before brokers start returning the new leaders.

Now let's change the concurrent map with the metadata cache. I believe the same logic still holds true i.e. an update metadata request containing the new topic identifiers needs to have propagated through the cluster before any requests using said topic identifiers can start to be serviced by the broker.

Am I wrong somewhere in my reasoning?

showuon commented:
@clolov , thanks for the detailed analysis. I admit I didn't check it so detailed last time. I just saw in `RLM#onLeadershipChange`, we did use `metadataCache`, so we should make sure `metadataCache` comes first before LISR. Had another look, we can still pass the `topicIDs` into `RLM#onLeadershipChange` to bypass it. So it looks fine.

For `fetchRemoteLogSegmentMetadata` and `findOffsetByTimestamp`, I have no concern since like you said, they are used in consumer, which means the metadataCache must be have updated. Although they can also be used by adminClient, some delay for the adminClient usage should be acceptable.

For `stopPartition`, I found not only Controller will trigger it, the client can send `deleteTopic` to trigger it. In this case, we might delete less partitions than expected like this case:
1. LISR arrived in RLM, created some RLM tasks for the leader partitions
2. delete topic comes in, miss the new added partitions in step (1)
3. updateMetadata arrived, updated the cache

I'm not quite sure if it's possible. TBH, I'm a little worried about introducing potential bugs due to some edge case we didn't know, due to we already confirmed there are some time gap between LISR and UMR. That's my thought. I'd also like to hear other reviewers' comments. 

Thanks again for the detailed analysis.

divijvaidya commented:
Hey folks

Let not target this change for 3.6 if we have doubts. It's anyways a perf improvement change. Let us get back to review this in detail after code freeze for 3.6 in another 2 weeks.

satishd commented:
Right. This is good to have refactoring and not blocking any feature for 3.6.

showuon commented:
> Let not target this change for 3.6 if we have doubts. It's anyways a perf improvement change. Let us get back to review this in detail after code freeze for 3.6 in another 2 weeks.

Agree.

> For stopPartition, I found not only Controller will trigger it, the client can send deleteTopic to trigger it.

I found I was wrong. It seems that the deleteTopic triggering `stopPartition` will only happen in KRaft, not ZK mode. But still unless we can 100% make sure there won't be any corner case, the time gap between LISR and UMR is still a concern to me.

Owen-CH-Leung commented:
Thanks all for your comments. I'll resolve conflict and work on this again after 3.6 is released

Owen-CH-Leung committed:
Merge branch 'trunk' into KAFKA-15038

Owen-CH-Leung committed:
Merge main and resolve conflict

Owen-CH-Leung commented:
I've just rebased trunk and resolved conflict. However when I run the test locally, I find that there're 2 failing tests at `kafka.admin.RemoteTopicCrudTest` on `testTopicDeletion`. 

I tried to debug further and find that when switch to metadatacache & when `stopPartitions` is called, `metadataCache.contains(tp)` is evaluated to `False` and thus the rest of the deletion logic is not processed. 
I believe this shouldn't be the case as the controller should update the metadataCache accordingly. Am I missing something that need to be changed ?

Owen-CH-Leung committed:
Merge branch 'trunk' into KAFKA-15038

Owen-CH-Leung committed:
merge trunk & Fix testTopicDeletion test

Owen-CH-Leung commented:
Hi all - I've rebased trunk. Can I ask for some feedback again ? Thanks

@clolov @showuon @satishd @kamalcph

kamalcph commented:
@Owen-CH-Leung 

Can you please rebase your PR against trunk? I'll review the PR in 2 days.

Owen-CH-Leung committed:
Merge branch 'trunk' into KAFKA-15038

Owen-CH-Leung committed:
merge trunk and resolve conflict

Owen-CH-Leung commented:
@kamalcph 

Thanks. Just rebased. 

Btw when I rebased, I notice that there's a failing test `kafka.server.AssignmentsManagerTest.testAssignmentAggregation` at trunk. Maybe we need to fix that also

kamalcph commented:
> RLM#onLeadershipChange, we did use metadataCache, so we should make sure metadataCache comes first before LISR. Had another look, we can still pass the topicIDs into RLM#onLeadershipChange to bypass it. So it looks fine.

We are still using metadata-cache in RLM#onLeadershipChange, we have to ensure that the `MetadataCache` is updated before handling the LeaderAndIsrRequest (LISR). Otherwise, we can follow the suggestion posted by @showuon. Also, there are existing test failures related to remote storage, can you please take a look?


```
Build / JDK 11 and Scala 2.13 / executeTieredStorageTest(String).quorum=kraft – org.apache.kafka.tiered.storage.integration.ReassignReplicaShrinkTest
Build / JDK 11 and Scala 2.13 / testSendOffsetsWithGroupMetadata(String).quorum=kraft – org.apache.kafka.tiered.storage.integration.TransactionsWithTieredStoreTest
Build / JDK 11 and Scala 2.13 / testBasicTransactions(String).quorum=kraft – org.apache.kafka.tiered.storage.integration.TransactionsWithTieredStoreTest
Build / JDK 11 and Scala 2.13 / testSendOffsetsWithGroupId(String).quorum=kraft – org.apache.kafka.tiered.storage.integration.TransactionsWithTieredStoreTest
Build / JDK 11 and Scala 2.13 / testDelayedFetchIncludesAbortedTransaction(String).quorum=kraft – org.apache.kafka.tiered.storage.integration.TransactionsWithTieredStoreTest
Build / JDK 21 and Scala 2.13 / testAbortTransactionTimeout(String).quorum=kraft – org.apache.kafka.tiered.storage.integration.TransactionsWithTieredStoreTest

```

Owen-CH-Leung committed:
Merge branch 'trunk' into KAFKA-15038

Owen-CH-Leung committed:
Update cache first before handling LISR

Owen-CH-Leung committed:
Merge branch 'trunk' into KAFKA-15038
